bigquery_etl/shredder/delete.py

```diff
@@ -392,36 +387,52 @@ def list_partitions(client, table, partition_expr, end_date, max_single_dml_byte
                 ).strip()
             ).result()
         }
-    with ThreadPoolExecutor(max_workers=args.parallelism) as executor:
-        results = await asyncio.gather(
-            *[
-                delete_from_table(
-                    client_q=client_q,
-                    executor=executor,
-                    target=replace(
-                        target, project=args.target_project or target.project
-                    ),
-                    source=replace(
-                        source, project=args.source_project or source.project
-                    ),
-                    source_condition=source_condition,
-                    dry_run=args.dry_run,
-                    read_only=args.read_only,
-                    priority=args.priority,
-                    start_date=args.start_date,
-                    end_date=args.end_date,
-                    max_single_dml_bytes=args.max_single_dml_bytes,
-                    state_table=args.state_table,
-                    states=states,
-                )
-                for target, source in DELETE_TARGETS.items()
-                if table_filter(target.table)
-            ]
+    tasks = [
+        task
+        for target, source in DELETE_TARGETS.items()
+        if table_filter(target.table)
+        for task in delete_from_table(
+            client=client,
+            target=replace(target, project=args.target_project or target.project),
+            source=replace(source, project=args.source_project or source.project),
+            source_condition=source_condition,
+            dry_run=args.dry_run,
+            read_only=args.read_only,
+            priority=args.priority,
+            start_date=args.start_date,
+            end_date=args.end_date,
+            max_single_dml_bytes=args.max_single_dml_bytes,
+            state_table=args.state_table,
+            states=states,
         )
-    if not results:
+    ]
+    if not tasks:
         logging.error("No tables selected")
         parser.exit(1)
-    bytes_processed, bytes_deleted = map(sum, zip(*results))
+    # ORDER BY partition_sort_key DESC, sql_table_id ASC
+    # https://docs.python.org/3/howto/sorting.html#sort-stability-and-complex-sorts
+    tasks.sort(key=lambda task: sql_table_id(task.table))
+    tasks.sort(key=attrgetter("partition_sort_key"), reverse=True)
```

Comment:
Am I understanding this correctly?

```
table_v1, 20200301
table_v1, 20200302
table_v2, 20200301
table_v2, 20200302
```

Would be sorted as:

```
table_v1, 20200302
table_v2, 20200302
table_v1, 20200301
table_v2, 20200301
```

The threadpool assigns tasks sequentially based on the ordering of the list, so DELETEs from the same table are unlikely to be scheduled at the same time, therefore increasing utilization of slots. 

---

Have your initial tests shown the expected increase in utilization?

Reply:
> Am I understanding this correctly?

yes

> Have your initial tests shown the expected increase in utilization?

yes